vault/identity_store_oidc_provider.go

```diff
@@ -684,3 +783,130 @@ func (i *IdentityStore) pathOIDCClientExistenceCheck(ctx context.Context, req *l
 
 	return entry != nil, nil
 }
+
+// pathOIDCCreateUpdateProvider is used to create a new named provider or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateProvider(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var provider provider
+	if req.Operation == logical.UpdateOperation {
+		entry, err := req.Storage.Get(ctx, providerPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&provider); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if issuerRaw, ok := d.GetOk("issuer"); ok {
+		provider.Issuer = issuerRaw.(string)
+	} else if req.Operation == logical.CreateOperation {
+		provider.Issuer = d.GetDefaultOrZero("issuer").(string)
+	}
+
+	if allowedClientIDsRaw, ok := d.GetOk("allowed_client_ids"); ok {
+		provider.AllowedClientIDs = allowedClientIDsRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		provider.AllowedClientIDs = d.GetDefaultOrZero("allowed_client_ids").([]string)
+	}
+
+	if scopesRaw, ok := d.GetOk("scopes"); ok {
+		provider.Scopes = scopesRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		provider.Scopes = d.GetDefaultOrZero("scopes").([]string)
+	}
+
+	scopeTemplateKeyNames := make(map[string]string)
+	for _, scopeName := range provider.Scopes {
+		entry, err := req.Storage.Get(ctx, scopePath+scopeName)
+		if err != nil {
+			return nil, err
+		}
+		// enforce scope existence on provider create and update
+		if entry == nil {
+			return logical.ErrorResponse("scope %q does not exist", scopeName), nil
+		}
+
+		// ensure no two templates have the same top-level keys
+		var storedScope scope
+		if err := entry.DecodeJSON(&storedScope); err != nil {
+			return nil, err
+		}
+
+		jsonTemplate := make(map[string]interface{})
+		if err = json.Unmarshal([]byte(storedScope.Template), &jsonTemplate); err != nil {
+			return nil, err
+		}
+
+		for keyName := range jsonTemplate {
+			val, ok := scopeTemplateKeyNames[keyName]
+			if ok && val != scopeName {
+				return logical.ErrorResponse("scope templates cannot have conflicting top-level keys; found conflict %q in scopes %q, %q", keyName, scopeName, val), nil
```

Comment:
As discussed, we'll change this to a warning on the response instead of returning an error.